### ISSUE TITLE
There is no CI

### ISSUE THREAD

**Issue by alanz** [lspitzner/brittany#36]
travis, circleci, appveyor, ...

**Comment by lspitzner** [comment]
is there a way to set up travis (or any other freely available one) to test `cabal build` `cabal new-build` and `stack build` ? the two cabal builds probably should produce the same result always, so it might be sufficient to do `cabal build` and `stack build` only (of which there are examples around, although it gets confusing again quickly because the multi-ghc-travis does not bother with stack..). But then "should produce the same result" is exactly the thing ci is supposed to confirm.

**Comment by lspitzner** [comment]
For the record, I simply [posted the above question to reddit](https://www.reddit.com/r/haskell/comments/6rr4ft/can_i_easily_set_up_github_ci_to_test_cabal_build/) and got some useful replies, although no existing approaches test all three. The stack template uses an approach for cabal that does not use caching in the way the multi-ghc-travis script does, but the latter is for cabal only.

Time to cook some fresh copy-pasta dish, I guess.

**Comment by lspitzner** [comment]
"progress report:"

I am testing a semi-handcrafted build+new-build+stack .travis.yml and my main issue currently is the following:

(tail of travis build log)
~~~~
Preprocessing library brittany-0.8.0.3...
[ 1 of 24] Compiling Language.Haskell.Brittany.Internal.Prelude ( src/Language/Haskell/Brittany/Internal/Prelude.hs, dist/build/Language/Haskell/Brittany/Internal/Prelude.o )
[ 2 of 24] Compiling Language.Haskell.Brittany.Internal.Config.Types ( src/Language/Haskell/Brittany/Internal/Config/Types.hs, dist/build/Language/Haskell/Brittany/Internal/Config/Types.o )


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
~~~~

I haven't tried adjusting the build configuration, because I'd really rather fix the underlying problem. I investigated a bit and found that compiling `Config/Types.hs` alone takes ghc 600MB of memory (max residency ~230MB) and ~35sec on my hardware. I suspect that the travis-build-servers don't have that amount of memory available and this stalls out the process to the 10min+ time.

Now this module is ~350 loc only, but a ton of derived instances. Bisecting the issue a bit:

(max residency/runtime, as reported by `ghc +RTS -s`; I incrementally cut more and more stuff in such a way that each step still compiles):

- 230MB/35sec original;
- 150MB/12sec without aeson instances;
- 80MB/1sec without any instance derivation or manual definitions (only 2 one-line functions and several plain data decls remaining);
- 50MB/1sec when removing the import of `import Data.Semigroup.Generic`;
- 20MB/<1sec when removing most of the other import statements;

Now maybe I am too demanding of the compiler, but only the last numbers look justifiable to me. A 30MB difference in ghc max. residency just by adding an import statement? (and yes I have checked, the code compiles without error before and after.)

Also, aeson seems to be a general memory/cpu hog during compilation not only of the package, but also when deriving instances.

**Comment by lspitzner** [comment]
Fixed in baac0a6a7e74ea70cae02c4d317e4a24bce8e9b8. Compile-times seem to be fluctuating a lot, between 10 to 50 minutes (for the same build, and in the presence of caching, i.e. the 50min time did not even include deps). And the upper limit is 50 minutes afaik, so it may easily run into this limit at some point..

I could try to reduce compilation time even more, but I really don't want to spend time on that. Over here the full project takes like 3min to compile, which is fine. So if this does not work properly, I'll disable again and live without CI.
